federationsender/queue/destinationqueue.go

```diff
@@ -47,6 +47,23 @@ type destinationQueue struct {
 	pendingPDUs        []*gomatrixserverlib.HeaderedEvent      // owned by backgroundSend
 	pendingEDUs        []*gomatrixserverlib.EDU                // owned by backgroundSend
 	pendingInvites     []*gomatrixserverlib.InviteV2Request    // owned by backgroundSend
+	retryServerCh      chan bool                               // interrupts backoff
+}
+
+// retry will clear the blacklist state and attempt to send built up events to the server,
+// resetting and interrupting any backoff timers.
+func (oq *destinationQueue) retry() {
+	// TODO: We don't send all events in the case where the server has been blacklisted as we
+	// drop events instead then. This means we will send the oldest N events (chan size, currently 128)
+	// and then skip ahead a lot which feels non-ideal but equally we can't persist thousands of events
+	// in-memory to maybe-send it one day. Ideally we would just shove these pending events in a database
+	// so we can send a lot of events.
+	oq.statistics.Success()
+	oq.retryServerCh <- true
```

Comment:
If the queue worker isn't actively backing off at this point, or it isn't running at all, then wouldn't this block indefinitely?

It might make sense to add another `atomic.Bool` alongside `running` (maybe call it `blocking` or something) and ensure that it's only `true` when we're *really* backing off.

Reply:
Hmm, yes this will block forever. I'll have a think.